pkg/adsclient/sotw/v3/client.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2022 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+// Package sotw provides an implementation of GRPC SoTW (State of The World) part of XDS client
+package sotw
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/golang/protobuf/ptypes/any"
+	status "google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	grpcStatus "google.golang.org/grpc/status"
+)
+
+var (
+	errInit    = errors.New("ads client: grpc connection is not initialized (use InitConnect() method to initialize connection)")
+	errNilResp = errors.New("ads client: nil response from xds management server")
+)
+
+type AdsClient interface {
+	// Initialize the gRPC connection with management server and send the initial Discovery Request.
+	InitConnect(clientConn grpc.ClientConnInterface, opts ...grpc.CallOption) error
+	// Fetch waits for a response from management server and returns response or error.
+	Fetch() (*Response, error)
+	// Ack acknowledge the validity of the last received response to management server.
+	Ack() error
+	// Nack acknowledge the invalidity of the last received response to management server.
+	Nack(message string) error
+}
+
+type Response struct {
+	Resources []*any.Any
+}
+
+type adsClient struct {
+	ctx     context.Context
+	mu      sync.RWMutex
+	nodeID  string
+	typeURL string
+
+	// streamClient is the ADS discovery client
+	streamClient discovery.AggregatedDiscoveryService_StreamAggregatedResourcesClient
+	// lastAckedResponse is the last response acked by the ADS client
+	lastAckedResponse *discovery.DiscoveryResponse
+	// lastReceivedResponse is the last response received from management server
+	lastReceivedResponse *discovery.DiscoveryResponse
+}
+
+func NewAdsClient(ctx context.Context, nodeID string, typeURL string) AdsClient {
+	return &adsClient{
+		ctx:     ctx,
+		nodeID:  nodeID,
+		typeURL: typeURL,
+	}
+}
+
+func (c *adsClient) InitConnect(clientConn grpc.ClientConnInterface, opts ...grpc.CallOption) error {
```

Comment:
Can you add doc strings to the exported methods on this struct?

Reply:
Added with https://github.com/envoyproxy/go-control-plane/commit/1ab595617393939429f91b8edaa479194b9c7654